File: tests/algorithms/x/PHP/networking_flow/ford_fulkerson.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  $INF = 1000000000;
  function breadth_first_search($graph, $source, $sink, &$parent) {
  global $INF;
  $visited = [];
  $i = 0;
  while ($i < count($graph)) {
  $visited = _append($visited, false);
  $i = $i + 1;
};
  $queue = [];
  $queue = _append($queue, $source);
  $visited[$source] = true;
  $head = 0;
  while ($head < count($queue)) {
  $u = $queue[$head];
  $head = $head + 1;
  $row = $graph[$u];
  $ind = 0;
  while ($ind < count($row)) {
  $capacity = $row[$ind];
  if ($visited[$ind] == false && $capacity > 0) {
  $queue = _append($queue, $ind);
  $visited[$ind] = true;
  $parent[$ind] = $u;
}
  $ind = $ind + 1;
};
};
  return $visited[$sink];
};
  function ford_fulkerson(&$graph, $source, $sink) {
  global $INF;
  $parent = [];
  $i = 0;
  while ($i < count($graph)) {
  $parent = _append($parent, -1);
  $i = $i + 1;
};
  $max_flow = 0;
  while (breadth_first_search($graph, $source, $sink, $parent)) {
  $path_flow = $INF;
  $s = $sink;
  while ($s != $source) {
  $prev = $parent[$s];
  $cap = $graph[$prev][$s];
  if ($cap < $path_flow) {
  $path_flow = $cap;
}
  $s = $prev;
};
  $max_flow = $max_flow + $path_flow;
  $v = $sink;
  while ($v != $source) {
  $u = $parent[$v];
  $graph[$u][$v] = $graph[$u][$v] - $path_flow;
  $graph[$v][$u] = $graph[$v][$u] + $path_flow;
  $v = $u;
};
  $j = 0;
  while ($j < count($parent)) {
  $parent[$j] = -1;
  $j = $j + 1;
};
};
  return $max_flow;
};
  $graph = [[0, 16, 13, 0, 0, 0], [0, 0, 10, 12, 0, 0], [0, 4, 0, 0, 14, 0], [0, 0, 9, 0, 0, 20], [0, 0, 0, 7, 0, 4], [0, 0, 0, 0, 0, 0]];
  echo rtrim(_str(ford_fulkerson($graph, 0, 5))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/x/Kotlin/networking_flow/ford_fulkerson.kt
fun <T> _listSet(lst: MutableList<T>, idx: Int, v: T) { while (lst.size <= idx) lst.add(v); lst[idx] = v }

fun _numToStr(v: Number): String {
    val d = v.toDouble()
    val i = d.toLong()
    return if (d == i.toDouble()) i.toString() else d.toString()
}

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

var INF: Int = (1000000000).toInt()
var graph: MutableList<MutableList<Int>> = mutableListOf(mutableListOf(0, 16, 13, 0, 0, 0), mutableListOf(0, 0, 10, 12, 0, 0), mutableListOf(0, 4, 0, 0, 14, 0), mutableListOf(0, 0, 9, 0, 0, 20), mutableListOf(0, 0, 0, 7, 0, 4), mutableListOf(0, 0, 0, 0, 0, 0))
fun breadth_first_search(graph: MutableList<MutableList<Int>>, source: Int, sink: Int, parent: MutableList<Int>): Boolean {
    var visited: MutableList<Boolean> = mutableListOf<Boolean>()
    var i: Int = (0).toInt()
    while (i < graph.size) {
        visited = run { val _tmp = visited.toMutableList(); _tmp.add(false); _tmp }
        i = i + 1
    }
    var queue: MutableList<Int> = mutableListOf<Int>()
    queue = run { val _tmp = queue.toMutableList(); _tmp.add(source); _tmp }
    _listSet(visited, source, true)
    var head: Int = (0).toInt()
    while (head < queue.size) {
        var u: Int = (queue[head]!!).toInt()
        head = head + 1
        var row: MutableList<Int> = graph[u]!!
        var ind: Int = (0).toInt()
        while (ind < row.size) {
            var capacity: Int = (row[ind]!!).toInt()
            if ((visited[ind]!! == false) && (capacity > 0)) {
                queue = run { val _tmp = queue.toMutableList(); _tmp.add(ind); _tmp }
                _listSet(visited, ind, true)
                _listSet(parent, ind, u)
            }
            ind = ind + 1
        }
    }
    return visited[sink]!!
}

fun ford_fulkerson(graph: MutableList<MutableList<Int>>, source: Int, sink: Int): Int {
    var parent: MutableList<Int> = mutableListOf<Int>()
    var i: Int = (0).toInt()
    while (i < graph.size) {
        parent = run { val _tmp = parent.toMutableList(); _tmp.add(0 - 1); _tmp }
        i = i + 1
    }
    var max_flow: Int = (0).toInt()
    while ((breadth_first_search(graph, source, sink, parent)) as Boolean) {
        var path_flow: Int = (INF).toInt()
        var s: Int = (sink).toInt()
        while (s != source) {
            var prev: Int = (parent[s]!!).toInt()
            var cap: Int = (((graph[prev]!!) as MutableList<Int>)[s]!!).toInt()
            if (cap < path_flow) {
                path_flow = cap
            }
            s = prev
        }
        max_flow = max_flow + path_flow
        var v: Int = (sink).toInt()
        while (v != source) {
            var u: Int = (parent[v]!!).toInt()
            _listSet(graph[u]!!, v, ((graph[u]!!) as MutableList<Int>)[v]!! - path_flow)
            _listSet(graph[v]!!, u, ((graph[v]!!) as MutableList<Int>)[u]!! + path_flow)
            v = u
        }
        var j: Int = (0).toInt()
        while (j < parent.size) {
            _listSet(parent, j, 0 - 1)
            j = j + 1
        }
    }
    return max_flow
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(_numToStr(ford_fulkerson(graph, 0, 5)))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
